### PR TITLE
[FlexAttention] add support for learnable biases in Inductor

### PR DESCRIPTION
**Pull request by drisspg**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #137452

# Summary

The follow up PR to: https://github.com/pytorch/pytorch/pull/137526.  In this pr, we actually update the lowerings for the flex_attention backwards kernel to generate fused backward gradient calculations for any captured buffers that require grads.

We are doing this using tl.atomic_add to scatter the correct gradients into zeroed out buffer for any captured buffers that required grads. Added many test cases and found.  Along the way found some masking bugs.

There are likely some performance cliffs here, specifically with D-types and on different GPUs. Planned to do this in a follow-up and profile the current strategy. We are explicitly choosing reduced memory over increased performance right now.

By using atomics, we do not need to realize a full attention scores matrix. However, this comes with two downsides. One, this is potentially slower in some cases, and two, the gradient calculation for any captured buffers is non-deterministic. 

## Worked Example

Lets do the case where you are reading from one bias that doesn't require grad and using this to index into another that does. 

ScoreMod:
```Python
bias = torch.randn(
    params.seq_length,
    device=self.device,
    dtype=params.dtype,
    requires_grad=True,
)

offset = torch.randint(
    0,
    params.seq_length,
    (params.seq_length,),
    device=self.device,
)

def score_mod(score, b, h, q_idx, kv_idx):
    return score + bias[offset[q_idx]]
    
```

I am removing all but the new subgraph injected into the backwards:

``` Python
    dsT = pT * (dpT - Di[None, :])
    # ~~~~~~~~~~~~~~~~~~~ Apply joint modification  ~~~~~~~~~~~~~~~~~~~
    grad_scores = (dsT)


    # ~~~~~~~~~~~~~~~~~~~ Apply other buffer grad writes ~~~~~~~~~~~~~
    idx_b = off_z
    idx_h = off_hq
    idx_m = m
    idx_n = n
    scatter_mask = offs_m1[None, :] < Q_LEN and offs_n1[:, None] < KV_LEN
    tmp4 = (dsT).to(tl.float32)
    tl.atomic_add(out_ptr1 + (tl.broadcast_to(tl.load(in_ptr16 + idx_m), tmp4.shape)), tmp4, scatter_mask, sem='relaxed')

    # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
## Key points
* We always accumulate to float 32 grad buffers regardless of the type in the forward. This is because we normally do all computation intra kernel w/ fp32 accumulation and we want the same behavior for atomic additions
* We are currently restricted to 1 scatter in the kenrel. I have some ideas on fx rewrites that would remove this restrictions but for now have nice error message w/ work around and will leave as a follow up.
* Will do more extensive performance/ memory profiling in a follow up.

### Toy E2E example
I have a toy E2E training example PR in the gym for now: https://github.com/pytorch-labs/attention-gym/pull/84/
I plan to update to a realistic learnable bias before landing 

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov @Chillee @yanboliang @BoyuanFeng

### CODE REVIEW

File: torch/_higher_order_ops/flex_attention.py [hunk 1]
```diff
@@ -847,6 +847,7 @@ def sdpa_dense_backward(
     actual_grad_query = torch.empty_like(query)
     actual_grad_key = torch.empty_like(key)
     actual_grad_value = torch.empty_like(value)
+    # TODO can this be based off if score_mod_other_buffers requires grad?
```
Comment (drisspg) [inline torch/_higher_order_ops/flex_attention.py:850]: remove

File: torch/_inductor/codegen/triton.py [hunk 2]
```diff
@@ -1536,6 +1537,12 @@ def indexing(
                 # indirect indexing
                 cse_var = self.cse.varname_map[var.name]
                 mask_vars.update(cse_var.mask_vars)
+            elif "(" in var.name:
```
Comment (drisspg) [inline torch/_inductor/codegen/triton.py:1540]: cc @Chillee If I can figure out how to strip out this leading `(` so I can actually match against symT template indexes I think we have good support for all the divisible cases

File: torch/_inductor/codegen/triton.py [hunk 3]
```diff
@@ -2067,6 +2073,8 @@ def store(
         elif mode is None:
             line = f"tl.store({var} + ({indexing.index_str}), {value}, {indexing.mask_str})"
         elif mode == "atomic_add":
+            if isinstance(index, sympy.core.symbol.Symbol) and "(idx_m)" == index.name:
```
Comment (drisspg) [inline torch/_inductor/codegen/triton.py:2076]: Need to figure out the proper way to broadcast, dont have a good way over injecting the copy_shape into indexing generically

File: torch/_inductor/kernel/flex_attention.py [hunk 4]
```diff
@@ -1933,8 +2071,10 @@ def flex_attention_backward(*args, **kwargs):
         grad_key = lowerings[aten.sum](broadcasted_grad_key, axis=0, keepdims=True)
         grad_value = lowerings[aten.sum](broadcasted_grad_value, axis=0, keepdims=True)
 
+    # import fbvscode; fbvscode.set_trace()
```
Comment (drisspg) [inline torch/_inductor/kernel/flex_attention.py:2074]: delete

File: torch/_inductor/select_algorithm.py [hunk 5]
```diff
@@ -463,17 +466,39 @@ def load(self, name: str, index: sympy.Expr):
...
                 ), f"Expected the subgraph to be a ComputedBuffer, got {type(subgraph)}"
-                if isinstance(subgraph.data, ir.InputBuffer):
+                # Handle scatter stores
+                if isinstance(subgraph, list):
+                    for scatter_graph in subgraph:
+                        assert isinstance(
+                            scatter_graph, ir.ComputedBuffer
+                        ), "Expected a scatter if subgraph is a list"
+
+                        def funx(x, strides):
```
Comment (drisspg) [inline torch/_inductor/select_algorithm.py:484]: I think this is a safe assumption in the divisible case but also naming

File: torch/_inductor/codegen/triton.py [hunk 6]
```diff
@@ -2036,6 +2040,15 @@ def store(
         elif mode is None:
             line = f"tl.store({var} + ({indexing.index_str}), {value}, {indexing.mask_str})"
         elif mode == "atomic_add":
+            if (
+                isinstance(index, sympy.core.symbol.Symbol)
+                and symbol_is_type(index, SymT.TEMPLATE_INDEX)
+                or "idx" in indexing.index_str
```
Comment (drisspg) [inline torch/_inductor/codegen/triton.py:2046]: We often have expressions of our index variables and therefore the symbol is type does not match correctly..

File: torch/_inductor/select_algorithm.py [hunk 7]
```diff
@@ -176,6 +176,47 @@ def finalize_all(self) -> str:
...
+        if "idx" in var:
+            var = var.strip("()")
+        return sympy_index_symbol(var)
+
+    def store(self, name, index, value, mode):
+        """Store value with input tracking.
+        This is needed for grads of captured buffers that need to be added as inputs to the kernel
+        """
+        self.kernel.args.input(name)
+        self.kernel.template_out = value
```
Comment (drisspg) [inline torch/_inductor/select_algorithm.py:208]: explain what this is doing

File: torch/_inductor/select_algorithm.py [hunk 8]
```diff
@@ -176,6 +177,63 @@ def finalize_all(self) -> str:
...
+            index_str = self._process_indexing(index)
+            var = self._add_kernel_input(name)
+            return f"tl.load({var} + {index_str})"
+        return f"({self.fixed_inputs[name]})"
+
+    def indirect_indexing(self, index_var: str, size, check, wrap_neg=True):
+        """Convert index variable to symbolic form."""
+        var = str(index_var)
+        # Match strings that are exactly of the form (idx_<word>)
+        if re.fullmatch(r"\(idx_\w+\)", var):
```
Comment (drisspg) [inline torch/_inductor/select_algorithm.py:208]: this is cursed

File: torch/_inductor/kernel/flex_attention.py [hunk 9]
```diff
@@ -93,10 +106,79 @@ def get_float32_precision():
         return "'tf32'"
 
 
+@dataclass(frozen=True)
+class SubgraphOutput:
+    """When building the subgraph buffers we fall into 1 of two cases:
+    1. The subgraph returns a single output node, this happens in the forward and
+        backward pass when there are no input_buffers that require grad
+    2. The subgraph has multiple outputs (i.e. input_buffers that require grad)
+        In this case we want to return these output_nodes from the lowering
```
Comment (drisspg) [inline torch/_inductor/kernel/flex_attention.py:115]: maybe update, since we need both the TensorBox to return as the output from the lowering but the compute buffer to actually inline into the kernel

File: torch/_inductor/kernel/flex_attention.py [hunk 10]
```diff
@@ -1768,16 +1899,42 @@ def flex_attention_backward(*args, **kwargs):
...
+    )
+    assert isinstance(all_joint_outputs, List)
+    assert (
+        all_joint_outputs[0] is not None
+    ), "joint_subgraph_buffer is None this is a bug!"
+    joint_subgraph_buffer = all_joint_outputs[0].compute_buffer
+
+    score_mod_other_buffer_grads = all_joint_outputs[len(joint_placeholder_inps) - 1 :]
+
+    score_mod_other_buffer_grads_compute = [
```
Comment (drisspg) [inline torch/_inductor/kernel/flex_attention.py:1927]: I used tree map sometimes but nother others cause it splats inner tuples but should probs comment

File: torch/_inductor/select_algorithm.py (hunk 8 above)
Comment (drisspg) [reply torch/_inductor/select_algorithm.py:208]: but legit seemed needed so idk...

File: torch/_inductor/select_algorithm.py [hunk 11]
```diff
@@ -410,69 +468,79 @@ def stride(self, name, index=None):
...
+
+    def _handle_scatter_graph(self, scatter_graph):
+        """Handle processing for a single scatter graph.
+
+        Args:
+            scatter_graph: The scatter graph to process
+        """
+        assert isinstance(
+            scatter_graph, ir.ComputedBuffer
+        ), "Expected a scatter if subgraph is a list"
```
Comment (drisspg) [inline torch/_inductor/select_algorithm.py:490]: this is weird error message

File: torch/_inductor/select_algorithm.py [hunk 12]
```diff
@@ -410,69 +468,79 @@ def stride(self, name, index=None):
...
+
+        def contiguous_strides(x):
```
Comment (drisspg) [inline torch/_inductor/select_algorithm.py:492]: would love confirmation on this

File: torch/_inductor/select_algorithm.py [hunk 13]
```diff
@@ -410,69 +468,79 @@ def stride(self, name, index=None):
...
-            with V.set_ops_handler(PlaceholderSubstitution(V.ops)):
+            subgraph = self._get_subgraph(subgraph_number)
+            modification_handler = ModificationWrapper(
+                self, subgraph_number, fixed_inputs, mask
+            )
+            with V.set_ops_handler(modification_handler):
                 assert isinstance(
-                    subgraph, ir.ComputedBuffer
+                    subgraph, (ir.ComputedBuffer, List)
                 ), f"Expected the subgraph to be a ComputedBuffer, got {type(subgraph)}"
```
Comment (drisspg) [inline torch/_inductor/select_algorithm.py:529]: message needs update

File: torch/_inductor/subgraph_lowering.py [hunk 14]
```diff
@@ -27,46 +41,89 @@ class PointwiseSubgraphLowering(torch.fx.Interpreter):
...
-            "Buffers cannot be created while lowering a pointwise subgraph. "
-            "This could be for a good reason (e.g. you're calling an op we can't codegen as a pointwise op), "
-            "but it could also be a bug. Please file a bug report if you think this should be supportable."
-        )
+    def register_buffer(self, buffer: ir.Buffer, *, set_name: bool = False) -> str:
+        if self._approved_mutator():
+            return self.root_graph.register_buffer(buffer, set_name=set_name)
+        else:
+            raise SubgraphLoweringException(
+                f"Buffer '{buffer.name}' cannot be registered during the lowering of {self._current_op}. "
```
Comment (drisspg) [inline torch/_inductor/subgraph_lowering.py:95]: This error message would be more helpful, but because Reduction operations are not necessarily the target that registers the buffer. The error message can be kind of confusing. For instance, for the existing captured buffer test, this is what we see.

targets that get looped over:
```
aten.sum.dim_IntList
aten.sqrt.default
aten.index.Tensor
```

`Buffers cannot be registered during the lowering of aten.index.Tensor. `

So although there is a reduction which breaks the pointwise lowering i.e the sum.dim list it's not until we actually access the elements that we have to register the new buffer. So maybe we just remove the current op from this error string since it can be kind of confusing.

File: torch/_inductor/select_algorithm.py (hunk 8 above)
Comment (Chillee) [inline torch/_inductor/select_algorithm.py:208]: ???

File: torch/_inductor/kernel/flex_attention.py [hunk 15]
```diff
@@ -1673,6 +1784,73 @@ def bwd_dkdv_block_mn(
...
+        ):
+            for user in node.users:
+                if user.op != "output":
+                    raise NotImplementedError(
+                        "Using multiple indexing operations on the same tensor that requires gradients "
+                        "in a score_mod function is not currently supported. "
+                        "This typically happens when indexing the same tensor multiple times, like:\n\n"
+                        "    def score_mod(score, b, h, q_idx, kv_idx):\n"
+                        "        return score + bias[q_idx] + bias[kv_idx]  # bias used twice!\n\n"
+                        "A valid workaround is to clone() the tensors that will be indexed multiple times, "
```
Comment (Chillee) [inline torch/_inductor/kernel/flex_attention.py:1801]: (i.e. `bias.clone()`)

File: torch/_inductor/kernel/flex_attention.py [hunk 16]
```diff
@@ -1891,8 +2083,17 @@ def flex_attention_backward(*args, **kwargs):
                 full_q_indices,
             ],
             layout=layout_broadcasted_k,  # We use store_output only for grad_key
-            subgraphs=[fw_subgraph_buffer, joint_subgraph_buffer, mask_graph_buffer],
-            mutated_inputs=[grad_query, broadcasted_grad_value],
+            subgraphs=[
+                fw_subgraph_buffer,
+                joint_outputs.joint_buffer,
```
Comment (Chillee) [inline torch/_inductor/kernel/flex_attention.py:2087]: Rename to something like `grad_score` (or `grad_inp`)

File: torch/_inductor/kernel/flex_attention.py [hunk 17]
```diff
@@ -1891,8 +2083,17 @@ def flex_attention_backward(*args, **kwargs):
...
+                mask_graph_buffer,
+                joint_outputs.buffer_grads_compute,
```
Comment (Chillee) [inline torch/_inductor/kernel/flex_attention.py:2089]: captured grads?

File: torch/_inductor/subgraph_lowering.py [hunk 18]
```diff
@@ -27,46 +41,89 @@ class PointwiseSubgraphLowering(torch.fx.Interpreter):
...
+    allowed_mutations: Optional[Set[OpOverload]] = None
+    additional_lowerings: Optional[LoweringDict] = None
+    _current_op: Optional[TargetType] = None
 
     def __init__(
         self,
         gm: torch.fx.GraphModule,
-        root_graph_lowering: "torch._inductor.graph.GraphLowering",
+        root_graph_lowering: torch._inductor.graph.GraphLowering,
+        allowed_mutations: Optional[Set[OpOverload]] = None,
```
[the remark this answers is not in the document]
Comment (drisspg) [reply torch/_inductor/subgraph_lowering.py:56]: talked offline Since buffers created via operators and we dont now a priori what the buffers are it makes sense to allow list via Op

**Review by drisspg — COMMENTED** [review]
,

**Review by Chillee — APPROVED** [review]

**Comment by Chillee** [comment]
!!
